features/0434-outofband/README.md

```diff
@@ -130,7 +130,57 @@ If only the `request~attach` item is included, no new connection is expected to
 
 If both the `handshake_protocols` and `request~attach` items are included in the message, the receiver should first establish a connection and then respond (using that connection) to one of the messages in the `request~attach` message. If a connection already exists between the parties, the receiver may respond immediately to the `request-attach` message using the established connection.
 
-While the _receiver_ is expected to respond with an initiating message from a `handshake_protocols` or `request~attach` item using an offered service, the receiver may be able to respond by reusing an existing connection. Specifically, if a connection they have was created from an out-of-band `invitation` from the same public DID of a new `invitation` message, the receiver **SHOULD** use the existing connection in responding to the `invitation`. If a selected `handshake_protocols` item is used by the receiver, it may have a specific "reuse this connection" message type.
+### Reuse Messages
+
+While the _receiver_ is expected to respond with an initiating message from a `handshake_protocols` or `request~attach` item using an offered service, the receiver may be able to respond by reusing an existing connection. Specifically, if a connection they have was created from an out-of-band `invitation` from the same public DID of a new `invitation` message, the receiver **MAY** use the existing connection in responding to the `invitation`. This is accomplished using the `reuse` and `reuse-accepted` messages.
```

Comment:
I think this should only be used if there is only a `handshake_protocols` entry and no `request~attach`.  If there is a `request~attach` (even if there is also a `handshake_protocols` item) and an existing connection, the first message back should be the response to the attachment, whatever that is, using the existing connection.

Reply:
In today's meeting, I agreed that we should indicate this, but I think we need to confirm one thing.

The `reuse-accepted` message handles an explicit case where the _inviter_ no longer has the keys to receive a message over the existing (known to the _invitee_). 

How should that be handled in this case? If the attached message is processed, and a message sent back to the _inviter_ using a key it no longer has, do we fall in the same hole?

Even though redundant in the common case, it may help to still send a reuse message prior to engaging via the attached messages to avoid falling into the same hole.

Thoughts? @swcurran

Reply:
@swcurran 

> I think this should only be used if there is only a handshake_protocols entry and no request~attach. If there is a request~attach (even if there is also a handshake_protocols item) and an existing connection, the first message back should be the response to the attachment, whatever that is, using the existing connection.

Why?

@TelegramSam 

> How should that be handled in this case? If the attached message is processed, and a message sent back to the inviter using a key it no longer has, do we fall in the same hole?

Yes.

> Even though redundant in the common case, it may help to still send a reuse message prior to engaging via the attached messages to avoid falling into the same hole.

Agreed.

Reply:
@llorllale -- about why this: "I think this should only be used if there is only a handshake_protocols entry and no `request~attach`. If there is a `request~attach` (even if there is also a handshake_protocols item) and an existing connection, the first message back should be the response to the attachment, whatever that is, using the existing connection."

If you get a handshake AND a request attach AND you already have a connection, why not just skip to the response to the attach, and not bother with the handshake?

Sam has raised the issue that if the invitation Sender has deleted the connection, they won't be able to process the request, but (a) that's likely an 80-20 rare case and (b) the worst that happens is the message is lost and eventually someone realizes it.  I think we should optimize for the common case.

I'm reluctantly OK with going through the reuse dance every time (two extra messages!!! The horror!!!  :-) ) But if that's what people want, we can do that.  I don't think it's the end of the world either way.

Reply:
BTW --- what do we do if we have a connection and there is ONLY the `request~attach` message?  We get into the same hole, but with no "reuse", "confirmed" message pair to fall back on.  Do we accept that as OK?

What about this approach:

- If the `request~attach` response in turn has a response coming back to the invitee, the reuse/confirmed message pair is really a waste, so we don't use it.
- If there is not a response expected to the `request~attach` response, we recommend the use of `~please_ack` with the `request~attach` response so that we get a response if all is well.

Between the two, we achieve the same result in all cases and bypass the extra messages reuse/confirmed.

@TelegramSam @llorllale